packages/compass-query-bar/src/components/query-options-grid.tsx

```diff
@@ -0,0 +1,227 @@
+  }
+
+  // When there are no numeric options, we want link to be last cell in
+  // last row.
+  if (numericEditors.length === 0) {
+    const lastDocumentEditor = documentEditors.pop();
+    const linkRow = `${cells(lastDocumentEditor!.name, COLUMNS - 1)} rest`;
+    const rows = [
+      documentEditors
+        // First row will have either 1 or 2 cells evenly distributed in 6
+        // available columns
+        .map((option) => {
+          return cells(option.name, COLUMNS / documentEditors.length);
+        })
+        .join(' '),
+      linkRow,
+    ];
+    return toTemplateArea(rows);
+  }
+
+  // For every other case we will do some special handling where numeric inputs
+  // will take 3 to 4 cells based on the amount of inputs provided and will
+  // always be on the end of the second row and document inputs adjust
+  // around that.
+  const numericCellsLength = numericEditors.length > 2 ? 4 : 3;
+  const freeCellsLength = COLUMNS - numericCellsLength;
+  const numericCells = cells('rest', numericCellsLength);
+
+  // If there is more than two editors in total, move last one to the next row.
+  if (documentEditors.length > 2) {
+    const lastDocumentEditor = documentEditors.pop();
+    const rows = [
+      documentEditors
+        .map((option) => {
+          return cells(option.name, COLUMNS / documentEditors.length);
+        })
+        .join(' '),
+      `${cells(lastDocumentEditor!.name, freeCellsLength)} ${numericCells}`,
+    ];
+    return toTemplateArea(rows);
+  }
+
+  // Otherwise keep numeric cells on the last row separately from document
+  // editors.
+  const rows = [
+    documentEditors
+      .map((option) => {
+        return cells(option.name, COLUMNS / documentEditors.length);
+      })
+      .join(' '),
+    `${cells('.', freeCellsLength)} ${numericCells}`,
+  ];
+  return toTemplateArea(rows);
+}
+
+type QueryOptionsGridProps = {
+  queryOptions: QueryOption[];
+  onApply: () => void;
+  onChangeQueryOption: (queryOption: QueryOption, value: string) => void;
+  queryOptionProps: QueryBarOptionProps;
+  refreshEditorAction: Listenable;
+  schemaFields: string[];
+  serverVersion: string;
+};
+
+export const QueryOptionsGrid: React.FunctionComponent<QueryOptionsGridProps> =
+  ({
+    queryOptions,
+    onApply,
+    onChangeQueryOption,
+    queryOptionProps,
+    refreshEditorAction,
+    schemaFields,
+    serverVersion,
+  }) => {
+    const gridTemplateAreas = useMemo(
+      () => getGridTemplateForQueryOptions(queryOptions),
+      [queryOptions]
+    );
+
+    const documentEditors = useMemo(
+      () =>
+        Object.values(OPTION_DEFINITION)
+          .filter((opt) => {
+            return (
+              opt.name !== 'filter' &&
+              queryOptions.includes(opt.name) &&
+              opt.type === 'document'
+            );
+          })
+          .map((optionDefinition) => optionDefinition.name),
+      [queryOptions]
+    );
+    const numericEditors = useMemo(
```

Comment:
We compute the numeric & document editors twice, here and in `getGridTemplateForQueryOptions`. I'm thinking I prefer keeping the argument interface of `getGridTemplateForQueryOptions` with only the query options array instead of passing these in for the slight reduction in duplication. I'm up to update it to reduce this duplication if folks prefer it.